patches/server/0817-Properly-remove-overriding-potion-effect-modifiers.patch

```diff
@@ -0,0 +1,190 @@
+attribute modifier added by the potion effect based on a single uuid
+defined by each potion effect type.
+After the removal, a new attribute modified is added that holds the
+updated modifications based on the lower amplifier of the weaker potion
+effect.
+As the removal of the stronger potion effect is purely based on the
+single uuid of each individual potion effect type the amplifier of the
+old effect is not needed to properly remove the effect and is sadly not
+passed to the removal logic in the first place.
+
+This however fails for the absorption potion effect type, as it cannot
+make use of the attribute modifier system and the single uuid concept
+used by other potion effect types.
+Hence, prior to this commit, it simply calculates the amount of
+absorption hearts granted by the amplifier of the potion effect and
+removes these.
+As the server however only provides the amplifier (more specifically the
+new potion effect) to the update method once an overriding absorption
+effect is replaced, the removal only removes the amount of absorption
+hearts.
+
+This patch combats this faulty behaviour by now properly passing the old
+and new mob effect (and their respective amplifier) through the update
+logic of the living entity and mob effects.
+While this prevents the absorption effect from leaving the player with
+absorption hearts after the effect completely expires, it did allow the
+effect to re-generate absorption hearts that were lost due to damage
+when replacing the overriding effect with the lower potion effect.
+
+The patch introduces MobEffect#updateAttributeModifiers which is
+responsible for performing the previously described logic of removing
+and applying attribute modifiers.
+This method is then leveraged by the absorption mob effect to correctly
+calculate the difference between the old and new amplifiers.
+
+Resolves: https://bugs.mojang.com/browse/MC-182497
+
+diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
+index b6f9c99d580d985f2b84efaa655d9ae40073e72b..3511ba42f0daf457eb4f5d7c98932d9fb1916479 100644
+--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
++++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
+@@ -1767,6 +1767,13 @@ public class ServerPlayer extends Player {
+     @Override
+     protected void onEffectUpdated(MobEffectInstance effect, boolean reapplyEffect, @Nullable Entity source) {
+         super.onEffectUpdated(effect, reapplyEffect, source);
++        // Paper start - properly remove old overridden potion effect
++    }
++
++    @Override
++    protected void onEffectUpdated(MobEffectInstance effect, @Nullable MobEffectInstance oldEffect, boolean reapplyEffect, @Nullable Entity source) {
++        super.onEffectUpdated(effect, oldEffect, reapplyEffect, source);
++        // Paper end - properly remove old overridden potion effect
```

Comment:
I this will break anywhere that calls the "legacy" onEffectUpdated method because its no longer gonna start tracking the levitation position. You should make the "legacy" method call the new one with a null extra argument.

Reply:
I am unsure if I understand exactly what you mean here.

The `levitationStartPos` is maintained by `ServerPlayer#onEffectAdded` which previously was also only called once (when the first levitation effect was added, updates did not re-trigger it). None of the changes in this patch touch the points at which `onEffectAdded` is called.

Additionally, the "legacy" method `ServerPlayer#onEffectUpdated` simply calls its super method `LivingEntity#onEffectUpdated`: 
```java
protected void onEffectUpdated(MobEffectInstance effect, boolean reapplyEffect, @Nullable Entity source) {
    // Paper start - properly remove old overridden potion effect
    this.onEffectUpdated(effect, null, reapplyEffect, source);
}
```
which will properly call the new `ServerPlayer#onEffectUpdated` with a null argument.

I might have completely misunderstood your point tho :sweat_smile: so if you could elaborate a bit on it, that would be great <3

Reply:
oh, I think I mis-typed, thinking I was looking at the effectsAdded method. What I should've said, was anything that called the legacy `ServerPlayer#onEffectUpdated` method would no longer trigger the advancement criteria or send a mob effect update packet, since that legacy method doesn't delegate to the overloaded method in ServerPlayer, rather the super method.

Reply:
Okay, sweet!